jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java

```diff
@@ -18,179 +18,394 @@
+ * </p>
+ *
+ * <code>
+ * HttpCompliance compliance = HttpCompliance.rfc7230Builder()
+ *     .allowCaseInsensitiveFieldNames(true)
+ *     .requireColonAfterFieldName(true)
+ *     .allowHttp09(false)
+ *     .allowMultipleContentLengths(true)
+ *     .build();
+ * </code>
  */
-public enum HttpCompliance // TODO in Jetty-10 convert this enum to a class so that extra custom modes can be defined dynamically
+public final class HttpCompliance
 {
-    /** A Legacy compliance mode to match jetty's behavior prior to RFC2616 and RFC7230. 
-     */
-    LEGACY(sectionsBySpec("0,METHOD_CASE_SENSITIVE")), 
-    
-    /** The legacy RFC2616 support, which incorrectly excludes 
-     * {@link HttpComplianceSection#METHOD_CASE_SENSITIVE}, 
-     * {@link HttpComplianceSection#FIELD_COLON},
-     * {@link HttpComplianceSection#TRANSFER_ENCODING_WITH_CONTENT_LENGTH},
-     * {@link HttpComplianceSection#MULTIPLE_CONTENT_LENGTHS},
+    /**
+     * Request attribute name for storing encountered compliance violations
      */
-    RFC2616_LEGACY(sectionsBySpec("RFC2616,-FIELD_COLON,-METHOD_CASE_SENSITIVE,-TRANSFER_ENCODING_WITH_CONTENT_LENGTH,-MULTIPLE_CONTENT_LENGTHS")), 
-    
-    /** The strict RFC2616 support mode */
-    RFC2616(sectionsBySpec("RFC2616")), 
-    
-    /** Jetty's current RFC7230 support, which incorrectly excludes  {@link HttpComplianceSection#METHOD_CASE_SENSITIVE} */
-    RFC7230_LEGACY(sectionsBySpec("RFC7230,-METHOD_CASE_SENSITIVE")),
-
-    /** The RFC7230 support mode */
-    RFC7230(sectionsBySpec("RFC7230")),
-    
-    /** Custom compliance mode that can be defined with System property <code>org.eclipse.jetty.http.HttpCompliance.CUSTOM0</code> */
-    @Deprecated
-    CUSTOM0(sectionsByProperty("CUSTOM0")),
-    /** Custom compliance mode that can be defined with System property <code>org.eclipse.jetty.http.HttpCompliance.CUSTOM1</code> */
-    @Deprecated
-    CUSTOM1(sectionsByProperty("CUSTOM1")),
-    /** Custom compliance mode that can be defined with System property <code>org.eclipse.jetty.http.HttpCompliance.CUSTOM2</code> */
-    @Deprecated
-    CUSTOM2(sectionsByProperty("CUSTOM2")),
-    /** Custom compliance mode that can be defined with System property <code>org.eclipse.jetty.http.HttpCompliance.CUSTOM3</code> */
-    @Deprecated
-    CUSTOM3(sectionsByProperty("CUSTOM3"));
-  
     public static final String VIOLATIONS_ATTR = "org.eclipse.jetty.http.compliance.violations";
 
-    private static final Logger LOG = Log.getLogger(HttpParser.class);
-    private static EnumSet<HttpComplianceSection> sectionsByProperty(String property)
+    // Identification of the mode
+    private String mode;
+    // The default values here are just Jetty defaults and not associated with any specific RFC/Spec
+    // Use the appropriate builder to start with a specific RFC/Spec behavior
+    private boolean allowCaseInsensitiveFieldNames = false;
```

Comment:
I think the data structure should be an `EnumSet<ViolationSection>` rather than multiple boolean fields that have to be kept in sync with the enum anyway.   By all means have convenience setters on the Builder like `allowMultiLineFieldValue(boolean)`, but they should just mutate the underlying `EnumSet`.

Reply:
Actually see my other comments elsewhere... `ViolationSection` should be named `ComplianceViolation` and the API should basically be:
```java
public final class HttpCompliance
{
    public static final HttpCompliance LEGACY = legacyBuilder().build();
    public static final HttpCompliance RFC2616 = rfc2616Builder().build();
    public static final HttpCompliance RFC7230 = rfc7230Builder().build();

    public static Builder builder() {...}
    public static Builder legacyBuilder() {...}
    public static Builder rfc2616Builder() {...}
    public static Builder rfc7230Builder() {...}

    public EnumSet<ComplianceViolation> getCompliance();

    public interface Builder()
    {
        void allow(ComplianceViolation violation);
        void disAllow(ComplanceViolation violation);
    }
}
```